### PR TITLE
EPMRPP-47547 Setup TypeScript

### PR DESCRIPTION
**Pull request by AmsterGet**

### CODE REVIEW

File: src/realTimeReporter/utils.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { EVENTS } from '../constants';
+
+export const publishEvent = (event: EVENTS, msg: any): void => {
+  // @ts-ignore
```
Comment (DiscoElevator) [inline src/realTimeReporter/utils.ts]: Will using `@types/node` help?

File: src/constants.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+export enum EVENTS {
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply src/constants.ts]: So you mean we have to create a separate file for each `enum`?

File: src/models/common.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import { STATUSES, LOG_LEVELS } from '../constants';
+
+export enum LAUNCH_MODES {
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply src/models/common.ts]: Ok

File: src/models/reporting.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+export interface FinishLaunchRQ extends RPItemFinishRQ {}
+
+export interface LogRQ extends RPLogRQ {}
+
+export interface AttachmentRQ extends RPLogRQ {
+    attachment: {
+        name: string,
+        type: FILE_TYPES,
+        content: any,
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply src/models/reporting.ts]: Fixed

File: src/realTimeReporter/utils.ts (hunk 1 above)
Comment (AmsterGet) [reply src/realTimeReporter/utils.ts]: Unfortunately no.
As we use methods from `process` implementation of `Process` interface and haven't such method overload.
image

Comment (AmsterGet) [reply src/realTimeReporter/utils.ts]: I found some solution. Please check it out (see the screenshot below).

image

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,20 @@
+{
+  "compilerOptions": {
+    "sourceMap": true,
+    "esModuleInterop": true,
+    "allowJs": false,
+    "noImplicitAny": true,
+    "moduleResolution": "node",
+    "lib": ["es2018"],
+    "module": "commonjs",
+    "target": "es2018",
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply tsconfig.json]: Fixed

File: src/realTimeReporter/reportingApi.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import { EVENTS, TEST_ITEM_TYPES } from '../constants';
+import { StartTestItemRQ, FinishTestItemRQ, LogRQ, AttachmentRQ, Attribute } from '../models';
+import { publishEvent } from './utils';
+
+export default class ReportingApi {
+
+  public static startSuite(data: StartTestItemRQ): void { // TODO: may be change it to common startTestItem and manage item type by browser object inside handler
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply src/realTimeReporter/reportingApi.ts]: I'll remove it in the PR with implementation of all listeners.

File: src/realTimeReporter/reporter.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
+import _ from 'lodash';
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply src/realTimeReporter/reporter.ts]: Removed.

File: package.json [hunk 8]
```diff
@@ -2,37 +2,43 @@
...
+    "lint": "eslint \"src/**/*.ts\"",
     "format": "npm run lint -- --fix",
-    "example:chai": "cd examples && npm run chai",
-    "example:chromeDriver": "cd examples && npm run chromeDriver",
-    "example:chromeDriver:realTimeReporter": "cd examples && npm run chromeDriver:realTimeReporter"
+    "example:chai": "npm run build && cd examples && npm run chai",
+    "example:chromeDriver": "npm run build && cd examples && npm run chromeDriver",
+    "example:chromeDriver:realTimeReporter": "npm run build && cd examples && npm run chromeDriver:realTimeReporter"
   },
   "dependencies": {
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply package.json]: It is already added [here](https://github.com/reportportal/agent-js-nightwatch/pull/8/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R47-R49)

File: src/constants.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (DiscoElevator) [reply src/constants.ts]: Yes

File: src/models/configs.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    token: string;
+    project: string;
+    endpoint: string;
+    launch: string;
+
+    debug?: boolean;
+    isLaunchMergeRequired?: boolean;
+}
+
+export interface AgentOptions {
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply src/models/configs.ts]: Fixed

File: src/constants.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (AmsterGet) [reply src/constants.ts]: Ok. Fixed

File: src/realTimeReporter/utils.ts (hunk 1 above)
Comment (DiscoElevator) [reply src/realTimeReporter/utils.ts]: Destructuring will break event emitter (functions will lose their context). Maybe it's not worth wasting too much time trying to make this perfectly typed.

Comment (AmsterGet) [reply src/realTimeReporter/utils.ts]: Agree. So I just leaved `@ts-ignore` here

**Comment by AmsterGet** [comment]
> It would be good to add some examples using TS, but I think it should be in the scope of another task.

Ok.

**Pull request merged**